### ISSUE TITLE
[HOLD for payment 2023-02-18] [$1000] Sending multiline messages with 1st line as h1 and 2nd line regular, both lines appears in the LHN 

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14670]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open the app
2. Open any chat
3. Send bold text using # and normal text on next line
4. Open another chat
5. Send same text as sent in point 3 but without making it bold using #

## Expected Result:
 Only first line of both the messages should be visible in the LHN

## Actual Result:
For chat with bold text and normal text on next line, both the bold text and normal text are visible adjacent to each other on LHN

## Workaround:
unknown;
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.61-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/215598841-d90b5be7-06a3-427a-9b19-23afdd0f36be.mp4

https://user-images.githubusercontent.com/43996225/215598857-c74a09ff-bd62-49a5-9597-53121ff29962.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1675068352846209

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0108354374b8e7c006</li>
        <li>Upwork Job ID: 1620446934113501184</li>
        <li>Last Price Increase: 2023-01-31</li>
    </ul>
</details>

**Comment by tienifr** [comment]
# Proposal

This is due to an issue in `parser.htmlToText` that we're parsing the `<h1>` incorrectly.

We're parsing things like 
`<h1>Test</h1>message` into `Testmessage`, which is incorrect because `h1` always make the next block go to next line. In order to fix this, we need to add a new rule to `htmlToTextRules` to parse `</h1>` into `/n` instead of empty string:

```diff
this.htmlToTextRules = [
            {
                name: 'breakline',
                regex: /<br[^>]*>/gi,
                replacement: '\n',
            },
            {
                name: 'blockquoteopen',
                regex: /(.|\s)(<blockquote>)/gi,
                replacement: '$1\n',
            },
            {
                name: 'blockquote',
                regex: /<\/blockquote>(.|\s)/gm,
                replacement: '\n$1',
            },
+            {
+                name: 'headingClose',
+                regex: /<\/h1>/gi,
+                replacement: '\n',
+            },
            {
                name: 'stripTag',
                regex: /(<([^>]+)>)/gi,
                replacement: '',
            },
        ];
```

**Comment by tienifr** [comment]
We can improve the regex for the `h1` if we want to, for example: `/<\/h1[^>]*>/gi` to account for the case when there's whitespace between `</h1` and `>`, but the basic idea is the same

**Comment by tienifr** [comment]
Working fine (for optimistic data case) after the fix:

https://user-images.githubusercontent.com/113963320/215822472-74426164-6dca-4a2b-94a3-c3c9ee804fe7.mov

**Comment by tienifr** [comment]
currently I see we're only using `h1`, not other headings, but if we want to fix for future cases of `h2`, `h3` we can do that in the regex too

**Comment by s77rt** [comment]
### Proposal
```diff
             {
-                name: 'blockquoteopen',
-                regex: /(.|\s)(<blockquote>)/gi,
+                name: 'opentag',
+                regex: /(.|\s)(<([^>]+)>)/gi,
                 replacement: '$1\n',
             },
             {
-                name: 'blockquote',
-                regex: /<\/blockquote>(.|\s)/gm,
+                name: 'closetag',
+                regex: /<\/([^>]+)>(.|\s)/gm,
                 replacement: '\n$1',
             },
```
Apply the changes [here](https://github.com/Expensify/expensify-common/blob/main/lib/ExpensiMark.js#L297-L306)

### RCA
`htmlToText` in `ExpensiMark` (expensify-common) does not add necessary newlines.

This issue affect not only heading tag (`h1`) but also many others such as `pre`.

Looks like we already fixed that for `blockquote` so I have just extended the regex to match every tag.


### Notes:
- If you apply the proposal by editing `node_modules` files, make sure to clear the cache then rebuild the app: `rm -rf node_modules/.cache && npm run web`
- In order to fix the bug completely , we must apply the same fix in the backend. To test the solution locally switch to offline mode first.

### Results

https://user-images.githubusercontent.com/16493223/215889739-1a66780a-9449-4ad2-b984-2e3beed52776.mp4

**Comment by tienifr** [comment]
# Proposal 2

Follow up from [here](https://github.com/Expensify/App/issues/14670#issuecomment-1410710907), with the similar approach as [here](https://github.com/Expensify/App/issues/14670#issuecomment-1410687878), for fixing elements with similar issues if we want to.

Basically any HTML block element will encounter this issue (including blockquote, h1-h6, div, p). The change below will fix all block elements:
(The full list of HTML block elements are taken from [here](https://www.w3schools.com/html/html_blocks.asp))

```diff
diff --git a/lib/ExpensiMark.js b/lib/ExpensiMark.js
index 7f94ffe..a9df9ac 100644
--- a/lib/ExpensiMark.js
+++ b/lib/ExpensiMark.js
@@ -295,14 +295,14 @@ export default class ExpensiMark {
                 replacement: '\n',
             },
             {
-                name: 'blockquoteopen',
-                regex: /(.|\s)(<blockquote>)/gi,
+                name: 'blockElementOpen',
+                regex: /(.|\s)(<(address|article|aside|blockquote|canvas|dd|div|dl|dt|fieldset|figcaption|figure|footer|form|h1|h2|h3|h4|h5|h6|header|hr|li|main|nav|noscript|ol|p|pre|section|table|tfoot|ul|video)>)/gi,
                 replacement: '$1\n',
             },
             {
-                name: 'blockquote',
-                regex: /<\/blockquote>(.|\s)/gm,
-                replacement: '\n$1',
+                name: 'blockElementClose',
+                regex: /<\/(address|article|aside|blockquote|canvas|dd|div|dl|dt|fieldset|figcaption|figure|footer|form|h1|h2|h3|h4|h5|h6|header|hr|li|main|nav|noscript|ol|p|pre|section|table|tfoot|ul|video)>(.|\s)/gm,
+                replacement: '\n$2',
             },
             {
                 name: 'stripTag',

```

Of course we can simplify the regex by limiting it to only the block elements that we actually use. I'd prefer to fix all to be future-proof.

**Comment by tienifr** [comment]
For the `blockElementOpen`, this regex might be better 
`/(.)<(address|article|aside|blockquote|canvas|dd|div|dl|dt|fieldset|figcaption|figure|footer|form|h1|h2|h3|h4|h5|h6|header|hr|li|main|nav|noscript|ol|p|pre|section|table|tfoot|ul|video)>/gi`
(notice the `(.)` in here compared to the existing `(.|\s)`)

By default, it the block element is already on a new line, HTML will not add any extra line before it. So when we convert the HTML to text, we might not want to add that extra line as well.

**Comment by Santhosh-Sellavel** [comment]
@puneetlath I think we should just handle this for only formatting/tags that we support now, what do you say?

**Comment by puneetlath** [comment]
Yeah, I agree. I think we don't want to assume that all elements would have a line break.

**Comment by tienifr** [comment]
@Santhosh-Sellavel  I think we should fix it for all the HTML block elements that we're currently using (pre, blockquote, h1 are the ones I found in the codebase)
If we just fix for `h1`, the issue will still be there for other block elements, for example [this issue](https://github.com/Expensify/App/issues/14719) where it occurs for `pre`

In any case, the list of elements in the regex [here](https://github.com/Expensify/App/issues/14670#issuecomment-1411376239) just needs to be narrowed down to the ones that we want to fix now

```diff
diff --git a/lib/ExpensiMark.js b/lib/ExpensiMark.js
index 7f94ffe..a9df9ac 100644
--- a/lib/ExpensiMark.js
+++ b/lib/ExpensiMark.js
@@ -295,14 +295,14 @@ export default class ExpensiMark {
                 replacement: '\n',
             },
             {
-                name: 'blockquoteopen',
-                regex: /(.|\s)(<blockquote>)/gi,
+                name: 'blockElementOpen',
+                regex: /(.|\s)<(blockquote|h1|pre)>/gi,
                 replacement: '$1\n',
             },
             {
-                name: 'blockquote',
-                regex: /<\/blockquote>(.|\s)/gm,
-                replacement: '\n$1',
+                name: 'blockElementClose',
+                regex: /<\/(blockquote|h1|pre)>(.|\s)/gm,
+                replacement: '\n$2',
             },
             {
                 name: 'stripTag',
```

**Comment by s77rt** [comment]
@Santhosh-Sellavel @puneetlath 
I think every tag element should be treated as if it produces a new line.
There is no case where we want to parse `<tag>textA</tag>textB` as `textAtextB`

**Comment by Santhosh-Sellavel** [comment]
@puneetlath 

@tienifr  [Proposal](https://github.com/Expensify/App/issues/14670#issuecomment-1413011449) Looks good, but the fix seems to be applied on the backend also, can you take care of that? 

Also, we should put this on hold till the backend code is fixed!

Let me know your thoughts, thanks!

**Comment by eh2077** [comment]
**Proposal** 

Based on @tienifr 's proposal, I'd like to propose making extra changes to the App's codebase.

The main idea is to format last message from `lastMessageHtml` instead of `lastMessageText`. Then we don't need to wait the fix from backend and don't need to fix backend data of sent messages.

See below diff
```diff
diff --git a/src/libs/OptionsListUtils.js b/src/libs/OptionsListUtils.js
index d88d1292a9..37ffbc8de3 100644
--- a/src/libs/OptionsListUtils.js
+++ b/src/libs/OptionsListUtils.js
@@ -327,7 +327,7 @@ function createOption(logins, personalDetails, report, reportActions = {}, {
         if (ReportUtils.isReportMessageAttachment({text: report.lastMessageText, html: report.lastMessageHtml})) {
             lastMessageTextFromReport = `[${Localize.translateLocal('common.attachment')}]`;
         } else {
-            lastMessageTextFromReport = Str.htmlDecode(report ? report.lastMessageText : '');
+            lastMessageTextFromReport = report ? ReportUtils.formatReportLastMessageHtml(report.lastMessageHtml) : '';
         }
 
         const lastActorDetails = personalDetailMap[report.lastActorEmail] || null;
diff --git a/src/libs/ReportUtils.js b/src/libs/ReportUtils.js
index cf7cee7d58..40609c2aa6 100644
--- a/src/libs/ReportUtils.js
+++ b/src/libs/ReportUtils.js
@@ -405,6 +405,17 @@ function formatReportLastMessageText(lastMessageText) {
         .substring(0, CONST.REPORT.LAST_MESSAGE_TEXT_MAX_LENGTH);
 }
 
+/**
+ * Convert the last message html to text and trim it to a fixed limit.
+ * @param {String} lastMessageHtml
+ * @returns {String}
+ */
+function formatReportLastMessageHtml(lastMessageHtml) {
+    const parser = new ExpensiMark();
+    const lastMessageText = parser.htmlToText(lastMessageHtml);
+    return formatReportLastMessageText(lastMessageText);
+}
+
 /**
  * Helper method to return a default avatar
  *
@@ -1276,6 +1287,7 @@ export {
     getIOUTotal,
     canShowReportRecipientLocalTime,
     formatReportLastMessageText,
+    formatReportLastMessageHtml,
     chatIncludesConcierge,
     isPolicyExpenseChat,
     getDefaultAvatar,
diff --git a/src/libs/SidebarUtils.js b/src/libs/SidebarUtils.js
index 9e6e1ed751..90424c7fcb 100644
--- a/src/libs/SidebarUtils.js
+++ b/src/libs/SidebarUtils.js
@@ -246,7 +246,7 @@ function getOptionData(reportID) {
     if (ReportUtils.isReportMessageAttachment({text: report.lastMessageText, html: report.lastMessageHtml})) {
         lastMessageTextFromReport = `[${Localize.translateLocal('common.attachment')}]`;
     } else {
-        lastMessageTextFromReport = Str.htmlDecode(report ? report.lastMessageText : '');
+        lastMessageTextFromReport = report ? ReportUtils.formatReportLastMessageHtml(report.lastMessageHtml) : '';
     }
 
     const lastActorDetails = personalDetails[report.lastActorEmail] || null;
diff --git a/src/libs/actions/Report.js b/src/libs/actions/Report.js
index 3b4f5468f4..f8a349013a 100644
--- a/src/libs/actions/Report.js
+++ b/src/libs/actions/Report.js
@@ -215,6 +215,7 @@ function addActions(reportID, text = '', file) {
     const optimisticReport = {
         lastActionCreated: currentTime,
         lastMessageText: ReportUtils.formatReportLastMessageText(lastAction.message[0].text),
+        lastMessageHtml: lastAction.message[0].html,
         lastActorEmail: currentUserEmail,
         lastReadTime: currentTime,
     };
diff --git a/src/pages/reportPropTypes.js b/src/pages/reportPropTypes.js
index 7251f7a776..2a8de9569f 100644
--- a/src/pages/reportPropTypes.js
+++ b/src/pages/reportPropTypes.js
@@ -31,6 +31,9 @@ export default PropTypes.shape({
     /** The text of the last message on the report */
     lastMessageText: PropTypes.string,
 
+    /** The html of the last message on the report */
+    lastMessageHtml: PropTypes.string,
+
     /** The time of the last message on the report */
     lastActionCreated: PropTypes.string,

```

**Comment by MelvinBot** [comment]
@puneetlath, @Santhosh-Sellavel Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by Santhosh-Sellavel** [comment]
@puneetlath bump https://github.com/Expensify/App/issues/14670#issuecomment-1414473743

**Comment by puneetlath** [comment]
Thanks for the bump. I like @tienifr's proposal [here](https://github.com/Expensify/App/issues/14670#issuecomment-1413011449) as well to fix it specifically for the current html elements. 

I'll take care of any back-end changes that are needed.

**Comment by MelvinBot** [comment]
📣 @tienifr You have been assigned to this job by @puneetlath!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by tienifr** [comment]
Thank for assigning me! The PR will be ready for review today.

**Comment by tienifr** [comment]
Hi @Julesssss The PR for this issue in expensify-common is ready for review. Please help me check it. Thank you!

**Comment by Julesssss** [comment]
Hey @tienifr, the Expensify-Common PR is merged and you should be able to use the latest version now.

**Comment by puneetlath** [comment]
Ok so next @tienifr we need an App PR for using the new version of Expensify-Common, is that right?

**Comment by tienifr** [comment]
@puneetlath @Julesssss I will create a PR ASAP. Thank you!

**Comment by tienifr** [comment]
Hi @techievivek , [the PR ](https://github.com/Expensify/App/pull/14940) is ready for review. Please help me check it. Thank you!

**Comment by tienifr** [comment]
@techievivek 

# Regression Test Proposal

**Bug**: Sending multiline messages with 1st line as h1 and 2nd line regular, both lines appears in the LHN

**Proposed Test Steps**:

1. Open the app
2. Turn off the internet (1)
3. Open any chat
4. Send bold text using # and normal text on next line
5. Verify that only first line of the messages is visible in the LHN

**(1) Because the solution has just been implemented in front-end side, we need to turn off the internet to prevent backend from updating wrong value. After updating BE side, we can skip step 2**
Do we :+1: or :-1:

**Comment by Santhosh-Sellavel** [comment]
@puneetlath Are backend changes being taken care of, when can we expect it to be live?

**Comment by techievivek** [comment]
Added a PR to update backend. @puneetlath I have requested your review on the PR.

**Comment by tienifr** [comment]
Hi @Santhosh-Sellavel, please help me check [the PR](https://github.com/Expensify/App/pull/14940). I have updated the test steps as you requested. Thank you!

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.69-2 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/14940


If no regressions arise, payment will be issued on 2023-02-18. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@Santhosh-Sellavel / @puneetlath] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@Santhosh-Sellavel / @puneetlath] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@Santhosh-Sellavel / @puneetlath] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@tienifr] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@puneetlath] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by puneetlath** [comment]
@Santhosh-Sellavel can you help with identifying the PR that caused the bug? 

@tienifr can you propose some regression test steps for this bug?

**Comment by Santhosh-Sellavel** [comment]
> @Santhosh-Sellavel can you help with identifying the PR that caused the bug

@puneetlath I think this more like a improvement to the feature, not a regression caused.

**Comment by puneetlath** [comment]
@dhanashree-sawant can you please apply to the Upwork job? https://www.upwork.com/jobs/~0108354374b8e7c006

**Comment by dhanashree-sawant** [comment]
Hi @puneetlath , thanks I have just applied.

**Comment by tienifr** [comment]
@puneetlath  @Santhosh-Sellavel 
I just proposed the regression test in (this comment](https://github.com/Expensify/App/issues/14670#issuecomment-1421985764)

> # Regression Test Proposal
>
> **Bug**: Sending multiline messages with 1st line as h1 and 2nd line regular, both lines appears in the LHN
> 
> **Proposed Test Steps**:
> 
> 1. Open the app
> 2. Turn off the internet (1)
> 3. Open any chat
> 4. Send bold text using # and normal text on next line
> 5. Verify that only first line of the messages is visible in the LHN
> 
> **(1) Because the solution has just been implemented in front-end side, we need to turn off the internet to prevent backend from updating wrong value. After updating BE side, we can skip step 2**
> Do we :+1: or :-1:

**Comment by puneetlath** [comment]
All done. Thanks everyone!
